src/hamcrest/core/assert_that.py

```diff
@@ -25,7 +26,7 @@ def assert_that(assertion: bool, reason: str = "") -> None:
     ...
 
 
-def assert_that(actual, matcher=None, reason=""):
+def assert_that(actual, matcher=_unassigned, reason=""):
```

Comment:
Can this perhaps be simplified? (Forgive my somewhat pseudocode.)

```python
# second parameter was a Matcher
if isinstance(matcher, Matcher):
    return _assert_match(actual, matcher, reason)

# first parameter was a Matcher, the caller maybe passed them in the wrong order? (existing warning)
if isinstance(actual, Matcher):
    warnings.warn('...')

# second parameter was not provided, third parameter may have been provided
if matcher is _unassigned:
    return _assert_bool(actual, reason)

if reason:
    warnings.warn('reason string will be ignored...')

# If first parameter was not a string, and second parameter was a string, assume truthiness check.
if isinstance(matcher, str):
    if not isinstance(actual, str):
        return _assert_bool(actual, matcher)

# If second parameter was not a string, the caller likely thought they were doing an equality check.
# If both parameters are strings, it is ambiguous whether the caller thought they were doing an equality check,
# or a truthiness check with a reason string. Err on the side of caution and log a warning.
# If a truthiness check was desired, they can resolve the warning by passing the reason as a named parameter.
warnings.warn('possible misuse...')
return _assert_bool(actual, matcher)
```

I think we should avoid introducing a `TypeError` at this moment, because that is a breaking change. It would be nicer to have an intermediate release that warns about suspicious usages, allowing people to fix these issues without having to fix everything at once.